mitiq/factories.py

```diff
@@ -0,0 +1,129 @@
+
+import mitiq.qiskit.qiskit_utils as qs_utils
+import numpy as np
+
+class Factory(object):
+    """
+    This object adaptively produces new noise scaling parameters based on a historical stack of previous noise
+    scale parameters and previously calculated expectation values. Noise scaling parameters are stored on the `in_stack`
+    and the running list of expectation values are stored on the `out_stack`.
+    """
+    def __init__(self, instack: List[float], outstack: List[float]) -> None:
+        self.instack = instack
+        self.outstack = outstack
+
+    def step(self, instack: List[float], outstack: List[float]) -> None:
+        raise NotImplementedError
+
+    def is_converged(self, instack: List[float], outstack: List[float]) -> None:
+        raise NotImplementedError
+
+    def reduce(self, expectations: List[float]) -> float:
+        raise NotImplementedError
+
+
+class BatchedFactory(Factory):
+    def __init__(self, scalars: List[float], instack: List[float] = None, outstack: List[float] = None) -> None:
+        """
+        Runs a series of scalar noise parameters serially.
+        :param scalars: List of scalar noise values to be executed.
+        :param instack: Running stack of noise values run so far.
+        :param outstack: Expectations calculated thus far.
+        """
+        if instack is None:
+            instack = []
+        if outstack is None:
+            outstack = []
+        super(BatchedFactory, self).__init__(instack, outstack)
+        self.scalars = scalars
+
+    def step(self, instack: List[float], outstack: List[float]) -> float:
+        try:
+            next_param = self.scalars[len(outstack)]
+        except IndexError:
+            raise IndexError(f"BatchedFactory cannot take another step. Number of batched scalars ({len(instack)}) "
+                             f"exceeded.")
+        self.instack.append(next_param)
+        return next_param
+
+    def is_converged(self, instack: List[float], outstack: List[float]) -> bool:
+        return len(instack) == len(self.scalars)
+
+# Specific extrapolation algorithms are given below.
+# They are implemented as child classes of Factory or BatchedFactory.
+
+# TODO list:
+# [x] Richardson's extrapolation
+# [x] Linear fit
+# [x] Polynomial fit
+# Log-Richardson's extrapolation
+# Log-Polynomial fit
+# Adaptive Linear fit
+# Adaptive Log-polinomial fit
+# Adaptive max-likelihood
+# Adaptive Bayesian
+# etc... 
+
+
+class RichardsonFactory(BatchedFactory):
+    """Factory object implementing the Richardson's extrapolation algorithm."""
+
+    def reduce(self, x: List[float], y: List[float]) -> float:
+        """ Given two lists of x and y values associated to an unknwn function y=f(x), returns 
+        the extrapolation of the function to the x=0 limit, i.e., an estimate of f(0).
+        The algorithm is based on the Richardson's extrapolation method.
+        """
+
+        def get_gammas(c: float) -> List[float]:
```

Comment:
I would define outside of `reduce` using `@staticmethod`.

Reply:
Ok, done! Not sure if, in the body of the class, static methods should be called with `self.reduce` or with `RichardsonFactory.reduce`. I used `self.reduce` but both seem to work.

Reply:
`get_gammas` is not necessary anymore. In the current version `PolyFactory.static_reduce` is used.